### PR TITLE
feat: Add initial project structure for QA Test Automation, including…

### PR DESCRIPTION
**Pull request by gramajoguadalupe**
## 📋 Description
Added comprehensive GitHub Actions CI/CD configuration and test framework improvements.

## 🚀 Changes Made
- ✅ Updated `cypress.config.js` with CI/CD specific settings
- ✅ Enhanced `package.json` with additional test scripts
- ✅ Added complete GitHub Actions workflows (ci.yml, nightly.yml, deploy-reports.yml)
- ✅ Configured multi-browser testing matrix
- ✅ Added API testing with Cypress and Newman
- ✅ Implemented accessibility testing
- ✅ Added security scanning
- ✅ Created comprehensive documentation

## 🧪 Testing
- [x] All existing tests pass
- [x] New scripts tested locally
- [x] CI/CD configuration validated

## 📈 Features Added
- Multi-browser testing (Chrome, Firefox, Edge)
- Separate test jobs for DemoQA and AutomationExercise
- Nightly regression testing with full matrix
- Automated report deployment to GitHub Pages
- Slack notifications integration

### CODE REVIEW

File: tests/e2e/automationexercise/api.spec.js [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+  describe('Products API', () => {
+    it('should get all products successfully', () => {
+      cy.request({
+        method: 'GET',
+        url: `${apiBaseUrl}/productsList`,
+        failOnStatusCode: false
+      }).then((response) => {
+        // Verify response status
+        expect(response.status).to.eq(200)
```
Comment (ulises-jeremias) [inline tests/e2e/automationexercise/api.spec.js:28]: Great response validation and error handling! Good use of `failOnStatusCode` for robust testing

File: tests/page-objects/automationexercise/home.page.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  
+  // Cart modal
+  get cartModal() { return cy.get('#cartModal') }
+  get continueShoppingButton() { return cy.get('.btn-success').contains('Continue Shopping') }
+  get viewCartButton() { return cy.get('a[href="/view_cart"]').contains('View Cart') }
+  
+  // Page methods
+  visit() {
+    cy.visit('/')
+    return this
```
Comment (ulises-jeremias) [inline tests/page-objects/automationexercise/home.page.js:81]: Excellent method chaining and fluent interface! Good separation of concerns

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+      - name: 🧪 Run Cypress API tests
+        run: npm run test:api
+        env:
+          CI: true
+          
+      - name: 📮 Run Newman/Postman tests
+        run: npm run test:newman
+        continue-on-error: true
+        
+      - name: 📊 Upload API test artifacts
```
Comment (ulises-jeremias) [inline .github/workflows/ci.yml:63]: Professional pipeline with proper artifact management!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            tests/videos/
+            tests/reports/
+            newman-results.json
+          retention-days: 7
+
+  e2e-tests:
+    name: 🌐 E2E Testing
+    runs-on: ubuntu-latest
+    needs: setup
+    strategy:
```
Comment (ulises-jeremias) [inline .github/workflows/ci.yml:79]: Good use of matrix strategy for browser testing!

File: tests/page-objects/automationexercise/home.page.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
+/**
+ * Home Page Object for AutomationExercise
+ * Contains all elements and methods for the home page
+ * @author Guada Gramajo
+ * @date 11-06-25
+ */
+
+class HomePage {
+  // Page elements
+  get logo() { return cy.get('img[src*="logo"]') }
```
Comment (ulises-jeremias) [inline tests/page-objects/automationexercise/home.page.js:10]: Is it possible to use data-testid attributes for more stable selectors?

File: tests/fixtures/automationexercise-test-data.json [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  "accessibility": {
+    "checks": {
+      "colorContrast": true,
+      "keyboardNavigation": true,
+      "altText": true,
+      "headingStructure": true,
+      "focusManagement": true
+    }
+  }
+}
```
Comment (ulises-jeremias) [inline tests/fixtures/automationexercise-test-data.json:187]: Consider adding more edge cases and boundary testing. Add data for negative testing scenarios

**Review by ulises-jeremias — COMMENTED** [review]
**Hey @gramajoguadalupe ! 👋**

Thank you so much for this incredible submission! 🙌 It reflects not only strong technical skills but also a clear understanding of what quality, maintainable, and scalable test automation looks like.

At NaNLABS, our goal is that everyone walks away with **something valuable to explore or reflect on**, so even in cases like this—where the bar is already incredibly high—we’ll still leave a few optional suggestions for growth or experimentation. These are not requirements but opportunities if you’re curious to take it even further 🚀

---

## 🌟 Highlights

Here are some of the things we absolutely loved:

* ✅ **Excellent architecture and folder structure** — clean, modular, and easy to follow
* ✅ **Comprehensive coverage** — UI, API, performance, accessibility, responsive behavior, and more
* ✅ **Professional CI/CD setup** — complete GitHub Actions implementation with multi-browser matrix
* ✅ **Advanced Cypress usage** — custom commands, well-structured Page Objects, fluent interface
* ✅ **Solid test data management** — thoughtful fixtures and organization
* ✅ **Great documentation** — detailed and easy to follow
* ✅ **Mindset focused on scalability and best practices** — your solution is future-proof and production-grade

---

## 💡 Suggestions (Totally Optional)

We’re including these only as exploratory opportunities. Your solution already goes above and beyond:

* 🔍 **Improve selector strategy** by using `data-testid` attributes for more robust and future-proof tests.
* 📊 **Introduce data-driven testing** to make your test suite even more flexible and scalable.
* 🧪 **Expand coverage** in DemoQA and explore additional flows like checkout and auth.
* ⚙️ **Handle edge cases** such as network failures, invalid data, or timeout scenarios.
* 🚦 **Explore Lighthouse or visual regression tools** for broader test coverage.
* 🔄 **Add a Base Page class** to reduce duplication and unify shared behaviors across Page Objects.

---

## ✅ Final Thoughts

This is an **outstanding submission** in every sense. The level of care, quality, and professionalism is evident throughout. If you continue evolving this project even further, it could serve as a reference repository for others learning test automation.

If you're interested in diving deeper into any of the areas we mentioned, we’re happy to share more ideas or resources.

Thanks again for the time, effort, and quality you’ve put into this challenge — truly impressive work 👏👏
